dlmalloc/src/malloc.c

```diff
@@ -5227,23 +5227,14 @@ static void try_init_allocator(void) {
 
   char *base = &__heap_base;
   // Try to use the linker pseudo-symbol `__heap_end` for the initial size of
-  // the heap, but if that's not defined due to LLVM being too old perhaps then
-  // round up `base` to the nearest `PAGESIZE`. The initial size of linear
-  // memory will be at least the heap base to this page boundary, and it's then
-  // assumed that the initial linear memory image was truncated at that point.
-  // While this reflects the default behavior of `wasm-ld` it is also possible
-  // for users to craft larger linear memories by passing options to extend
-  // beyond this threshold. In this situation the memory will not be used for
-  // dlmalloc.
-  //
-  // Note that `sbrk(0)`, or in dlmalloc-ese `CALL_MORECORE(0)`, is specifically
-  // not used here. That captures the current size of the heap but is only
-  // correct if the we're the first to try to grow the heap. If the heap has
-  // grown elsewhere, such as a different allocator in place, then this would
-  // incorrectly claim such memroy as our own.
+  // the heap.
   char *end = &__heap_end;
-  if (end == NULL)
-    end = (char*) page_align((size_t) base);
+  if (end == NULL) {
+    // This can happen when 1. you are using an old wasm-ld which doesn't
```

Comment:
Wouldn't this just fail with `undefined symbol: __heap_end` on older versions of llvm?

If `foo` is not weak I'm pretty sure `&foo` can never be null.

Reply:
usually yes.
but see the comment in the code.

Reply:
But I don't see why `&__heap_end` could ever be null here.. that would be linker bug I think.    

I'm not sure why you mean by "something (other libraries or maybe provide a weak reference to `__heap_end`" ... do you mean that this symbol is weakly defined elsewhere?  If so then it should have a non-zero address still.

Reply:
In other words I don't think its possible for anyone to ever provide a symbol which has a null address.    

The only way its address can be null is if the reference here in this file is weak.

Reply:
> But I don't see why `&__heap_end` could ever be null here.. that would be linker bug I think.
> 
> I'm not sure why you mean by "something (other libraries or maybe provide a weak reference to `__heap_end`" ... do you mean that this symbol is weakly defined elsewhere?

yes.

eg. have `extern __attribute__((weak)) unsigned char __heap_end` in your app.

> If so then it should have a non-zero address still.

i observed NULL address when i experimented.
i don't know if it's a linker bug or not.

Reply:
If having a second, weak, reference to the same symbol causes the strong reference to get a zero address that would certainly be a linker bug.  The behaviour should be that if the program contains any strong references to a missing symbol then the link should fail.

Reply:
ok.
i wrote a small demo. https://github.com/yamt/garbage/tree/master/c/weak

Reply:
and filed a llvm issue: https://github.com/llvm/llvm-project/issues/60829